src/OpenEventSourcing.RabbitMQ/Connections/RabbitMqConnection.cs

```diff
@@ -0,0 +1,306 @@
+                channel.ExchangeDeclarePassive(exchange);
+            }
+            catch (OperationInterruptedException)
+            {
+                // NOTE(Dan): If we get here is means that the exhange doesn't exist.
+                throw new ExchangeNotFoundException(queue);
+            }
+
+            try
+            {
+                channel.QueueDeclarePassive(queue);
+            }
+            catch (OperationInterruptedException)
+            {
+                // NOTE(Dan): If we get here is means that the queue doesn't exist.
+                throw new QueueNotFoundException(queue);
+            }
+
+            channel.QueueUnbind(queue, exchange, filter);
+
+            ReturnChannel(channel);
+        }
+
+        public async Task<bool> ExchangeExistsAsync(string name)
+        {
+            await Task.Yield();
+
+            var channel = CreateChannel();
+
+            try
+            {
+                channel.ExchangeDeclarePassive(name);
+
+                return true;
+            }
+            catch (OperationInterruptedException)
+            {
+                return false;
+            }
+            finally
+            {
+                ReturnChannel(channel);
+            }
+        }
+        public async Task<bool> QueueExistsAsync(string name)
+        {
+            await Task.Yield();
+
+            var channel = CreateChannel();
+
+            try
+            {
+                channel.QueueDeclarePassive(name);
+
+                return true;
+            }
+            catch (OperationInterruptedException)
+            {
+                return false;
+            }
+            finally
+            {
+                ReturnChannel(channel);
+            }
+        }
+
+        public void Dispose()
+        {
+            if (_disposed)
+                return;
+
+            _disposed = true;
+
+            _pool.ReturnConnection(this, reRegisterForFinalization: false);
+        }
+        public void Reset()
+        {
+            _disposed = false;
+        }
+
+        private IModel CreateChannel()
+        {
+            if (_channels.IsEmpty)
+                return UnderlyingConnection.CreateModel();
+            
+            if (!_channels.TryDequeue(out var model))
+                return UnderlyingConnection.CreateModel();
+
+            if (model == null || !model.IsOpen)
+                return UnderlyingConnection.CreateModel();
+
+            return model;
+        }
+        private void ReturnChannel(IModel channel)
+        {
+            if (channel.IsOpen)
+            {
+                _channels.Enqueue(channel);
+                return;
+            }
+
+            channel.Dispose();
+        }
+
+        ~RabbitMqConnection()
+        {
+            _pool?.ReturnConnection(this, reRegisterForFinalization: true);
+        }
+
+        // TODO(Dan): Deal with the underlying connection events: `ConnectionShutdown`, `CallbackException` and `ConnectionBlocked`
```

Comment:
Does this need to be done for the first release?

Reply:
Was planning to get an initial package released, then deal with these afterwards. Want to be able to consume the package and start testing in more real world samples/products to get feedback on the API / usage etc